src/trio/_core/_tests/test_ki.py

```diff
@@ -517,3 +523,111 @@ async def inner() -> None:
         _core.run(inner)
     finally:
         threading._active[thread.ident] = original  # type: ignore[attr-defined]
+
+
+_T = TypeVar("_T")
+
+
+def _identity(v: _T) -> _T:
+    return v
+
+
+@pytest.mark.xfail(
+    strict=True,
+    raises=AssertionError,
+    reason=(
+        "it was decided not to protect against this case, see discussion in: "
+        "https://github.com/python-trio/trio/pull/3110#discussion_r1802123644"
+    ),
+)
+async def test_ki_does_not_leak_accross_different_calls_to_inner_functions() -> None:
+    assert not _core.currently_ki_protected()
+
+    def factory(enabled: bool) -> Callable[[], bool]:
+        @_identity(_core.enable_ki_protection if enabled else _identity)
+        def decorated() -> bool:
+            return _core.currently_ki_protected()
+
+        return decorated
```

Comment:
Why not call the conditional function on this to avoid the `_identity` dance

Reply:
Arbitrary choice really. I want to use it as a decorator

But now I really want to use it, so I can use pep 614